### PR TITLE
fix implementation of Function#chain

### PR DESCRIPTION
**Pull request by davidchambers**
See #1936

/cc @benperez

### CODE REVIEW

File: test/chain.js [hunk 1]
```diff
@@ -33,7 +33,7 @@ describe('chain', function() {
         return r + a;
       };
     };
-    var bound = R.chain(h, f);
+    var bound = R.chain(f, h);
     // (>>=) :: (r -> a) -> (a -> r -> b) -> (r -> b)
     // h >>= f = \w -> f (h w) w
     eq(bound(10), (10 * 2) + 10);
```
Comment (buzzdecafe) [inline test/chain.js]: i wouldn't mind seeing the `chain(append, head)([1, 2, 3]) //=> [1, 2, 3, 1]` example tested here. For my money that is a very clear illustration of the behavior

Comment (davidchambers) [inline test/chain.js]: :zap:

**Comment by buzzdecafe** [comment]
i'm glad i inadvertently stumbled upon the problem that resulted in this fix

:dromedary_camel:

File: src/chain.js [hunk 2]
```diff
@@ -24,12 +24,12 @@ var map = require('./map');
...
  *      R.chain(duplicate, [1, 2, 3]); //=> [1, 1, 2, 2, 3, 3]
+ *
+ *      R.chain(R.append, R.head)([1, 2, 3]); //=> [1, 2, 3, 1]
  */
 module.exports = _curry2(_dispatchable('chain', _xchain, function chain(fn, monad) {
   if (typeof monad === 'function') {
-    return function() {
-      return monad.call(this, fn.apply(this, arguments)).apply(this, arguments);
-    };
+    return function(x) { return fn(monad(x))(x); };
```
Comment (buzzdecafe) [inline src/chain.js]: of all the constraints we try to satisfy, supporting `this` context is the first one i would throw overboard.

Comment (davidchambers) [inline src/chain.js]: Down with `this`!

**Comment by benperez** [comment]
Nice catch folks, the implementation doesn't match the signature so that was definitely a bug. I think it probably went uncaught because we were trying to do something crazy like this 😛 :
`monad.call(this, fn.apply(this, arguments)).apply(this, arguments);`

How do we feel about the fact that our `Chain` method takes its args in the opposite order of `Fantasy-Land` (and also from Haskell's `(>>=)` FWIW). In the general case, I think taking the bound function first (`R.chain(x => fnThatReturnsAMaybe, Just(1))` makes more sense, since that's what we'd want to partially apply in a point-free style.

**Comment by buzzdecafe** [comment]
> How do we feel about the fact that our Chain method takes its args in the opposite order of Fantasy-Land

Seems consistent with `map`, e.g.

FantasyLand `map`
`map :: Functor f => f a ~> (a -> b) -> f b`

Ramda `map`
`map :: Functor f =>  (a -> b) -> f a -> f b`

FantasyLand `chain`
`chain :: Chain m => m a ~> (a -> m b) -> m b`

Ramda `chain`
`chain :: Chain m => (a -> m b) -> m a -> m b`

**Comment by davidchambers** [comment]
Adding `ap` to the picture makes the similarities between the functions even more apparent:

```haskell
map   :: Functor f => (a -> b) -> f a -> f b
ap    :: Apply f => f (a -> b) -> f a -> f b
chain :: Chain f => (a -> f b) -> f a -> f b
```

**Comment by benperez** [comment]
Cool, 🐪 from me on this change

**Comment by buzzdecafe** [comment]
me too
🐪

**Comment by CrossEye** [comment]
:herb:

**Pull request merged**